src/sc/sidechain.cpp

```diff
@@ -503,4 +506,128 @@ bool CSidechain::GetCeasingCumTreeHash(CFieldElement& ceasedBlockCum) const
     return true;
 }
 
+#if 0
+static int getInitNumBlocksForScFeeCheck()
+#else
+int CSidechain::getNumBlocksForScFeeCheck()
+{
+    if ( (Params().NetworkIDString() == "regtest") )
+    {
+        int val = (int)(GetArg("-blocksforscfeecheck", Params().ScNumBlocksForScFeeCheck() ));
+        if (val >= 0)
+        {
+            LogPrint("sc", "%s():%d - %s: using val %d \n", __func__, __LINE__, Params().NetworkIDString(), val);
+            return val;
+        }
+        LogPrint("sc", "%s():%d - %s: val %d is negative, using default %d\n",
+            __func__, __LINE__, Params().NetworkIDString(), val, Params().ScNumBlocksForScFeeCheck());
+    }
+    return Params().ScNumBlocksForScFeeCheck();
+}
+#endif
+
+#if 0
+int CSidechain::getNumBlocksForScFeeCheck()
+{
+    // gets constructed just one time
+    static int retVal( getInitNumBlocksForScFeeCheck() );
+    return retVal;
+}
+#endif
+
+void CSidechain::InitScFees()
+{
+    // only in the very first time, calculate the size of the buffer
+    if (sizeOfScFeesContainers == -1)
+    {
+        int epochLength = fixedParams.withdrawalEpochLength;
+
+        assert(epochLength > 0);
+       
+        int numBlocks = getNumBlocksForScFeeCheck();
+        sizeOfScFeesContainers = numBlocks / epochLength;
+        if (sizeOfScFeesContainers == 0 || (numBlocks % epochLength != 0) )
+        {
+            sizeOfScFeesContainers++;
+        }
+        LogPrint("sc", "%s():%d - sizeOfScFeesContainers set to %d\n", __func__, __LINE__, sizeOfScFeesContainers);
+
+        assert(sizeOfScFeesContainers > 0);
+        assert(scFees.empty());
+        Sidechain::ScFeeData defaultData(
+            lastTopQualityCertView.forwardTransferScFee, lastTopQualityCertView.mainchainBackwardTransferRequestScFee);
+        scFees.resize(sizeOfScFeesContainers, defaultData);
+    }
+}
+
+void CSidechain::UpdateScFees(const CScCertificateView& certView)
+{
+    InitScFees();
+
+    CAmount ftScFee   = certView.forwardTransferScFee;
+    CAmount mbtrScFee = certView.mainchainBackwardTransferRequestScFee;
+
+    // check if we are past the buffer size
+    if (scFees.size() >= sizeOfScFeesContainers)
```

Comment:
Consider the case with `sizeOfScFeesContainers` decrement if in regtest node was restarted with the flag `-blocksforscfeecheck` less than before, it may need to remove more than one element.

Reply:
Ok, done